### DISCUSSION TITLE
关于下载任务的referer跟useragent

### DISCUSSION THREAD

**Discussion by jc3213** [jc3213/download_with_aria2#26] [Ideas]
我把`referer`跟`user-agent`从`header`里分离出来了，原因是参考了某个提取百度网盘的脚本，那个脚本得到的结果如下
![Baidu](https://user-images.githubusercontent.com/8744936/156871052-67bd7b7d-9aa6-44e1-a3fe-2efa179f7c48.png)

因此我修改了原版跟多档案版的代码，我这边暂时没机会测试Firefox版本，只能测试Chrome版本貌似没什么问题？

**Comment by jc3213** [comment]
@ivysrono @daiaji

**Reply by ivysrono** [reply]
问题可能出在哪？我基本不下百度网盘。

**Reply by jc3213** [reply]
最开始是生成
```javascript
options.headers = ["Cookie: xxxx; Referer: yyyy; User-Agent: zzzz"]
```
现在是生成
```javascript
options.referer = yyyy
options.headers = ["Cookie: xxxx; "]
options['user-agent'] = zzzz
````
但是我火狐那边只是单纯改了代码没机会测试。如果可以的话帮忙试一下是否会出现错误（主要是看这个），比如输入时打错字符啥的导致问题之类。

**Reply by daiaji** [reply]
UA很正常，就是图标上的角标显示E。
请求里Cookie为空，正常吗？

**Reply by daiaji** [reply]
```json
{"id":"","jsonrpc":2,"method":"aria2.addUri","params":["token:454545454545",["https://download.manjaro.org/xfce/21.2.4/manjaro-xfce-21.2.4-220301-linux515.iso"],{"referer":"https://manjaro.org/downloads/official/xfce/","header":["Cookie:"],"user-agent":"Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/97.0.4692.71 Safari/537.36","all-proxy":""}]}
```

**Reply by jc3213** [reply]
> ```json
> {"id":"","jsonrpc":2,"method":"aria2.addUri","params":["token:454545454545",["https://download.manjaro.org/xfce/21.2.4/manjaro-xfce-21.2.4-220301-linux515.iso"],{"referer":"https://manjaro.org/downloads/official/xfce/","header":["Cookie:"],"user-agent":"Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/97.0.4692.71 Safari/537.36","all-proxy":""}]}
> ```

如果是网站本身就没cookie的话应该就是空的。主要是测试一些网盘（比如邮箱附件啥的）

**Reply by daiaji** [reply]
下载没问题，但是捕获下载的时候，浏览器自带的下载器的气泡会显示这个文件的读条。

**Reply by jc3213** [reply]
> 下载没问题，但是捕获下载的时候，浏览器自带的下载器的气泡会显示这个文件的读条。

这个是`downloads` API本身的问题，火狐的`onCreated`事件是在下载开始以后才触发。所以我才考虑引入`webRequest` API

Chrome那边`onCreated`跟火狐的`webRequest`差不多，直接就抓取了。`onDeterminingFilename`事件则是弹出确定文件名（另存为，如果是自动保存的话则是得到文件名那一步）的时候触发。不管哪一个都是在下载开始之前。

**Reply by ivysrono** [reply]
测了下，网易邮箱和北大法宝没问题。

**Reply by jc3213** [reply]
没问题就好
